flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -0,0 +1,609 @@
+			reportSlotStatus(taskExecutorConnection.getInstanceID(), initialSlotReport);
+			return false;
+		} else {
+			if (!taskExecutorManager.registerTaskManager(taskExecutorConnection, initialSlotReport)) {
+				LOG.debug("Task executor {} could not be registered.", taskExecutorConnection.getResourceID());
+				return false;
+			}
+
+			// register the new slots
+			for (SlotStatus slotStatus : initialSlotReport) {
+				slotTracker.addSlot(
+					slotStatus.getSlotID(),
+					slotStatus.getResourceProfile(),
+					taskExecutorConnection,
+					slotStatus.getJobID());
+			}
+
+			checkResourceRequirements();
+			return true;
+		}
+	}
+
+	@Override
+	public boolean unregisterTaskManager(InstanceID instanceId, Exception cause) {
+		checkInit();
+
+		LOG.debug("Unregistering task executor {} from the slot manager.", instanceId);
+
+		if (taskExecutorManager.isTaskManagerRegistered(instanceId)) {
+			slotTracker.removeSlots(taskExecutorManager.getSlotsOf(instanceId));
+			taskExecutorManager.unregisterTaskExecutor(instanceId);
+			checkResourceRequirements();
+
+			return true;
+		} else {
+			LOG.debug("There is no task executor registered with instance ID {}. Ignoring this message.", instanceId);
+
+			return false;
+		}
+	}
+
+	/**
+	 * Reports the current slot allocations for a task manager identified by the given instance id.
+	 *
+	 * @param instanceId identifying the task manager for which to report the slot status
+	 * @param slotReport containing the status for all of its slots
+	 * @return true if the slot status has been updated successfully, otherwise false
+	 */
+	@Override
+	public boolean reportSlotStatus(InstanceID instanceId, SlotReport slotReport) {
+		checkInit();
+
+		LOG.debug("Received slot report from instance {}: {}.", instanceId, slotReport);
+
+		if (taskExecutorManager.isTaskManagerRegistered(instanceId)) {
+			slotTracker.notifySlotStatus(slotReport);
+			checkResourceRequirements();
+			return true;
+		} else {
+			LOG.debug("Received slot report for unknown task manager with instance id {}. Ignoring this report.", instanceId);
+
+			return false;
+		}
+	}
+
+	/**
+	 * Free the given slot from the given allocation. If the slot is still allocated by the given
+	 * allocation id, then the slot will be marked as free and will be subject to new slot requests.
+	 *
+	 * @param slotId identifying the slot to free
+	 * @param allocationId with which the slot is presumably allocated
+	 */
+	@Override
+	public void freeSlot(SlotID slotId, AllocationID allocationId) {
+		checkInit();
+		LOG.debug("Freeing slot {}.", slotId);
+
+		slotTracker.notifyFree(slotId);
+		checkResourceRequirements();
+	}
+
+	// ---------------------------------------------------------------------------------------------
+	// Requirement matching
+	// ---------------------------------------------------------------------------------------------
+
+	private void checkResourceRequirements() {
```

Comment:
We should state the computational complexity of this method and add a comment that we might need to revisit this logic in case that the matching takes too much time. An alternative approach would be to remember which pending slot was assigned to which job. This could decrease the complexity a bit.